appdaemontestframework/appdaemon_mock/__init__.py

```diff
@@ -0,0 +1,39 @@
+import asyncio
+import appdaemon.utils
+
+# Appdaemon 4 uses Python asyncio programming. Since our tests are not async
+# we replace the sync_wrapper decorator with one that will always result in
+# synchronizing appdatemon calls.
+def sync_wrapper(coro):
+    def inner_sync_wrapper(self, *args, **kwargs):
+        start_new_loop = None
+        # try to get the running loop
+        # `get_running_loop()` is new in Python 3.7, fall back on privateinternal for 3.6
+        try:
+            get_running_loop = asyncio.get_running_loop
+        except AttributeError:
+            get_running_loop = asyncio._get_running_loop
+
+        # If there is no running loop we will need to start a new one and run it to completion
+        try:
+            if get_running_loop():
+                start_new_loop = False
+            else:
+                start_new_loop = True
+        except RuntimeError:
+            start_new_loop = True
+
+        if start_new_loop is True:
+            f = asyncio.ensure_future(coro(self, *args, **kwargs))
+            asyncio.get_event_loop().run_until_complete(f)
+            f = f.result()
+        else:
+            # don't use create_task. It's python3.7 only
+            f = asyncio.ensure_future(coro(self, *args, **kwargs))
+
+        return f
+
+    return inner_sync_wrapper
```

Comment:
This is my first exposure to `asynio` as well, so I'm far from an expert. This framework within Python seems to be rapidly evolving and having breaking changes making even harder to work with. I'll give my best shot at my understanding of things.

The appdeamon version will detect if we are called from a async context and only attempt to synchronize the call if it isn't. In the test framework, we currently only support synchronous operation, and the wrapper will always synchronize the call. Due to the way pytest sets up the environment, this required a little trickery to avoid deadlocks. I don't remember 100% of the details from when I originally wrote this. Let me know if you think it's worth re-hashing this and adding some deeper documentation around it.

Reply:
Thanks for your explanation.

It's still not 100% clear to me what's happening here, in the details I mean, and I'm not too comfortable adding a bit of code that I don't understand to the project. That being said. . . . the test coverage is so fantastic, and this seems to work. So I'm not going to stall the PR any longer just for this. 

However, understand that if this start breaking down, I will most likely ping you on slack to ask you to help me figure it out 😅